core/drivers/versal_huk.c

```diff
@@ -0,0 +1,339 @@
+#include <trace.h>
+#include <utee_defines.h>
+
+static struct {
+	uint8_t key[HW_UNIQUE_KEY_LENGTH];
+	bool ready;
+} huk;
+
+#define MODULE_SHIFT 8
+#define MODULE_ID 5
+#define API_ID(__x) ((MODULE_ID << MODULE_SHIFT) | (__x))
+
+#define	VERSAL_AES_KEY_SIZE_256  2
+#define VERSAL_AES_GCM_ENCRYPT 0
+
+enum versal_aes_key_src {
+	VERSAL_AES_BBRAM_KEY = 0,
+	VERSAL_AES_BBRAM_RED_KEY,
+	VERSAL_AES_BH_KEY,
+	VERSAL_AES_BH_RED_KEY,
+	VERSAL_AES_EFUSE_KEY,
+	VERSAL_AES_EFUSE_RED_KEY,
+	VERSAL_AES_EFUSE_USER_KEY_0,
+	VERSAL_AES_EFUSE_USER_KEY_1,
+	VERSAL_AES_EFUSE_USER_RED_KEY_0,
+	VERSAL_AES_EFUSE_USER_RED_KEY_1,
+	VERSAL_AES_KUP_KEY,
+	VERSAL_AES_PUF_KEY,
+	VERSAL_AES_USER_KEY_0,
+	VERSAL_AES_USER_KEY_1,
+	VERSAL_AES_USER_KEY_2,
+	VERSAL_AES_USER_KEY_3,
+	VERSAL_AES_USER_KEY_4,
+	VERSAL_AES_USER_KEY_5,
+	VERSAL_AES_USER_KEY_6,
+	VERSAL_AES_USER_KEY_7,
+	VERSAL_AES_EXPANDED_KEYS,
+	VERSAL_AES_ALL_KEYS,
+};
+
+enum versal_crypto_api {
+	VERSAL_AES_INIT = 96U,
+	VERSAL_AES_OP_INIT,
+	VERSAL_AES_UPDATE_AAD,
+	VERSAL_AES_ENCRYPT_UPDATE,
+	VERSAL_AES_ENCRYPT_FINAL,
+	VERSAL_AES_DECRYPT_UPDATE,
+	VERSAL_AES_DECRYPT_FINAL,
+	VERSAL_AES_KEY_ZERO,
+	VERSAL_AES_WRITE_KEY,
+	VERSAL_AES_LOCK_USER_KEY,
+	VERSAL_AES_KEK_DECRYPT,
+	VERSAL_AES_SET_DPA_CM,
+	VERSAL_AES_DECRYPT_KAT,
+	VERSAL_AES_DECRYPT_CM_KAT,
+};
+
+struct versal_aes_input_param {
+	uint64_t input_addr;
+	uint32_t input_len;
+	uint32_t is_last;
+};
+
+struct versal_aes_init {
+	uint64_t iv_addr;
+	uint32_t operation;
+	uint32_t key_src;
+	uint32_t key_len;
+};
+
+/*
+ * The PLM is little endian. When programming the keys in uint32_t the driver
+ * will BE swap the values.
+ *
+ * This way the test key below corresponds to the byte array 0xf8, 0x78, 0xb8,
+ * 0x38, 0xd8, 0x58, 0x98, 0x18, 0xe8, 0x68, ....
+ *
+ * NOTICE: This hardcoded value in DEVEL_KEY could have just been zeroes as done
+ * in the weak implementation found in otp_stubs.c.
+ */
+#define DEVEL_KEY { \
+	0xf878b838, 0xd8589818, 0xe868a828, 0xc8488808, \
+	0xf070b030, 0xd0509010, 0xe060a020, 0xc0408000, }
+
+#define AAD { \
+	0x67, 0xe2, 0x1c, 0xf3, 0xcb, 0x29, 0xe0, 0xdc, 0xbc, 0x4d, 0x8b, \
+	0x1d, 0x0c, 0xc5, 0x33, 0x4b, }
+
+#define NONCE { \
+	0xd2, 0x45, 0x0e, 0x07, 0xea, 0x5d, 0xe0, 0x42, 0x6c, 0x0f, 0xa1, \
+	0x33, }
+
+static bool versal_persistent_key(enum versal_aes_key_src src, bool *secure)
+{
+	struct versal_efuse_puf_sec_ctrl_bits puf_ctrl = { };
+	struct versal_efuse_sec_ctrl_bits ctrl = { };
+	struct versal_puf_data puf_data = { };
+	struct versal_puf_cfg cfg = {
+		.global_var_filter = VERSAL_PUF_GLBL_VAR_FLTR_OPTION,
+		.read_option = VERSAL_PUF_READ_FROM_EFUSE_CACHE,
+		.puf_operation = VERSAL_PUF_REGEN_ON_DEMAND,
+		.shutter_value = VERSAL_PUF_SHUTTER_VALUE,
+		.reg_mode = VERSAL_PUF_SYNDROME_MODE_4K,
+	};
+
+	switch (src) {
+	case VERSAL_AES_EFUSE_USER_KEY_0:
+		if (versal_efuse_read_sec_ctrl(&ctrl))
+			panic();
+
+		if (ctrl.user_key0_wr_lk)
+			*secure = true;
+		else
+			*secure = false;
```

Comment:
I think the compiler will do the right thing with just `*secure = ctrl.user_key0_wr_lk;`.